File: examples/C/rtmama.c
//
//  Custom routing Router to Mama (ROUTER to REQ)
//
#include "zhelpers.h"
#include <pthread.h>

#define NBR_WORKERS 10

static void *
worker_task (void *args)
{
    void *context = zmq_init (1);
    void *worker = zmq_socket (context, ZMQ_REQ);

    //  We use a string identity for ease here
    s_set_id (worker);
    zmq_connect (worker, "ipc://routing.ipc");

    int total = 0;
    while (1) {
        //  Tell the router we're ready for work
        s_send (worker, "ready");

        //  Get workload from router, until finished
        char *workload = s_recv (worker);
        int finished = (strcmp (workload, "END") == 0);
        free (workload);
        if (finished) {
            printf ("Processed: %d tasks\n", total);
            break;
        }
        total++;

        //  Do some random work
        s_sleep (randof (1000) + 1);
    }
    zmq_close (worker);
    zmq_term (context);
    return NULL;
}

//  .split main task
//  While this example runs in a single process, that is just to make
//  it easier to start and stop the example. Each thread has its own
//  context and conceptually acts as a separate process.

int main (void)
{
    void *context = zmq_init (1);
    void *client = zmq_socket (context, ZMQ_ROUTER);
    zmq_bind (client, "ipc://routing.ipc");
    srandom ((unsigned) time (NULL));

    int worker_nbr;
    for (worker_nbr = 0; worker_nbr < NBR_WORKERS; worker_nbr++) {
        pthread_t worker;
        pthread_create (&worker, NULL, worker_task, NULL);
    }
    int task_nbr;
    for (task_nbr = 0; task_nbr < NBR_WORKERS * 10; task_nbr++) {
        //  LRU worker is next waiting in queue
        char *address = s_recv (client);
        char *empty = s_recv (client);
        free (empty);
        char *ready = s_recv (client);
        free (ready);

        s_sendmore (client, address);
        s_sendmore (client, "");
        s_send (client, "This is the workload");
        free (address);
    }
    //  Now ask mamas to shut down and report their results
    for (worker_nbr = 0; worker_nbr < NBR_WORKERS; worker_nbr++) {
        char *address = s_recv (client);
        char *empty = s_recv (client);
        free (empty);
        char *ready = s_recv (client);
        free (ready);

        s_sendmore (client, address);
        s_sendmore (client, "");
        s_send (client, "END");
        free (address);
    }
    zmq_close (client);
    zmq_term (context);
    return 0;
}

File: examples/Ruby/rtmama.rb
# Custom routing Router to Mama (ROUTER to REQ)
# Ruby version, based on the C version.
#
# While this example runs in a single process, that is just to make
# it easier to start and stop the example. Each thread has its own
# context and conceptually acts as a separate process.
#
# libzmq: 2.1.10
# ruby: 1.9.2p180 (2011-02-18 revision 30909) [i686-linux]
# ffi-rzmq: 0.9.0
#
# @author Pavel Mitin
# @email mitin.pavel@gmail.com

require 'rubygems'
require 'ffi-rzmq'

WORKER_NUMBER = 10

def receive_string(socket)
  result = ''
  socket.recv_string result
  result
end

def worker_task
  context = ZMQ::Context.new 1
  worker = context.socket ZMQ::REQ
  #  We use a string identity for ease here
  worker.setsockopt ZMQ::IDENTITY, sprintf("%04X-%04X", rand(10000), rand(10000))
  worker.connect 'ipc://routing.ipc'

  total = 0
  loop do
    # Tell the router we're ready for work
    worker.send_string 'ready'

    # Get workload from router, until finished
    workload = receive_string worker

    p "Processed: #{total} tasks" and break if workload == 'END'
    total += 1

    # Do some random work
    sleep((rand(10) + 1) / 10.0)
  end
end

context = ZMQ::Context.new 1
client = context.socket ZMQ::ROUTER
client.bind 'ipc://routing.ipc'

workers = (1..WORKER_NUMBER).map do
  Thread.new { worker_task }
end

(WORKER_NUMBER * 10).times do
  # LRU worker is next waitin in queue
  address = receive_string client
  empty = receive_string client
  ready = receive_string client

  client.send_string address, ZMQ::SNDMORE
  client.send_string '', ZMQ::SNDMORE
  client.send_string 'This is the workload'
end

# Now ask mamas to shut down and report their results
WORKER_NUMBER.times do
  address = receive_string client
  empty = receive_string client
  ready = receive_string client

  client.send_string address, ZMQ::SNDMORE
  client.send_string '', ZMQ::SNDMORE
  client.send_string 'END'
end

workers.each &:join